k8scrdclient/k8scrdclient_v1.go

```diff
@@ -0,0 +1,288 @@
+			return microerror.Mask(err)
+		}
+
+		// In case the CRDs names are not accepted we have to stop processing here
+		// and return the reason of the failing condition. Therefore we stop retries
+		// permanently.
+		{
+			con, ok := statusConditionV1(manifest.Status.Conditions, apiextensionsv1.NamesAccepted)
+			if ok && statusConditionFalseV1(con) {
+				return backoff.Permanent(microerror.Maskf(nameConflictError, con.Message))
+			}
+		}
+		// In case the CRD is non-structural we have to stop processing here and
+		// return the reason of the failing condition. Therefore we stop retries
+		// permanently.
+		{
+			con, ok := statusConditionV1(manifest.Status.Conditions, apiextensionsv1.NonStructuralSchema)
+			if ok && statusConditionTrueV1(con) {
+				return backoff.Permanent(microerror.Maskf(notEstablishedError, con.Message))
+			}
+		}
+		// In case the CRD is not yet established we have to retry and only return a
+		// normal error so that the backoff can do its job.
+		{
+			con, ok := statusConditionV1(manifest.Status.Conditions, apiextensionsv1.Established)
+			if ok && statusConditionFalseV1(con) {
+				return microerror.Maskf(notEstablishedError, con.Message)
+			}
+		}
+
+		return nil
+	}
+
+	err = backoff.Retry(o, b)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+// crdVersionLatest returns true when the desired version of the given CRD
+// represents the latest version of them all. In case the current version of the
+// given CRD which we just read from the system appears to be newer, this means
+// another version of the same operator updated the CRD to its own version,
+// because it was in fact the latest version. Consider the following
+// relationships between operators and CRD versions they are aware of. Below o1
+// is operator-1 being aware of apiversion v1. In turn operator-2 knows about
+// apiversion v1 and v2, where v1 must be the exact same version as operator-1
+// defined already in order to keep the CRD schema version lifecycle in tact.
+// Note that in the example below o1 must not run anymore in cae o3 drops its
+// support. There is some orchestration involved the system's maintainers have
+// to get right.
+//
+//         o1    |     o2     |    o3
+//     ---------------------------------
+//         v1    |   v1, v2   |   v2, v3
+//
+// Each operator compares the latest version it finds in desired and current. If
+// the latest version of desired is below the latest version of current, the
+// desired version of the given CRD is not considered to be the latest known
+// version and thus not allowed to update the system's CRD.
+//
+// Note that the given versions must be in the format of usual Kubernetes
+// APIVersions, e.g. v1alpha1, v2beta5, v2.
+//
+//     https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning
+//
+func crdVersionLatestV1(desired *apiextensionsv1.CustomResourceDefinition, current *apiextensionsv1.CustomResourceDefinition) (bool, error) {
```

Comment:
Do we need this now with https://github.com/giantswarm/opsctl/pull/596 ?

Reply:
We need to talk about how to approach the current situation. I just tried to figure out what actually works and what does not.